hathor/feature_activation/feature_service.py

```diff
@@ -0,0 +1,170 @@
+        _block = _Block.from_block(block)
+        offset_to_boundary = _block.height % self._settings.evaluation_interval
+        previous_boundary_height = _block.height - (offset_to_boundary or self._settings.evaluation_interval)
+        previous_boundary_block = _get_ancestor_at_height(_block=_block, height=previous_boundary_height)
+        previous_state = self.get_state(block=previous_boundary_block, feature=feature)
+
+        if offset_to_boundary != 0:
+            return previous_state
+
+        return self._calculate_new_state(
+            boundary_block=_block,
+            feature=feature,
+            previous_state=previous_state
+        )
+
+    def _calculate_new_state(
+        self,
+        *,
+        boundary_block: _Block,
+        feature: Feature,
+        previous_state: FeatureState
+    ) -> FeatureState:
+        """Returns the new feature state based on the new block, the criteria, and the previous state."""
+        assert boundary_block.height % self._settings.evaluation_interval == 0, (
+            'cannot calculate new state for a non-boundary block'
+        )
+        criteria = self._get_criteria(feature=feature)
+
+        if previous_state is FeatureState.DEFINED:
+            if boundary_block.height >= criteria.start_height:
+                return FeatureState.STARTED
+
+            return FeatureState.DEFINED
+
+        if previous_state is FeatureState.STARTED:
+            if boundary_block.height >= criteria.timeout_height and not criteria.activate_on_timeout:
+                return FeatureState.FAILED
+
+            if (
+                boundary_block.height >= criteria.timeout_height
+                and criteria.activate_on_timeout
+                and boundary_block.height >= criteria.minimum_activation_height
+            ):
+                return FeatureState.ACTIVE
+
+            count = self.get_bit_count(boundary_block=boundary_block.data, bit=criteria.bit)
+            threshold = criteria.threshold if criteria.threshold is not None else self._settings.default_threshold
+
+            if (
+                boundary_block.height < criteria.timeout_height
+                and count >= threshold
+                and boundary_block.height >= criteria.minimum_activation_height
+            ):
+                return FeatureState.ACTIVE
+
+            return FeatureState.STARTED
+
+        if previous_state is FeatureState.ACTIVE:
+            return FeatureState.ACTIVE
+
+        if previous_state is FeatureState.FAILED:
+            return FeatureState.FAILED
+
+        raise ValueError(f'Unknown previous state: {previous_state}')
+
+    def _get_criteria(self, *, feature: Feature) -> Criteria:
+        criteria = self._settings.features.get(feature)
+
+        if not criteria:
+            raise ValueError(f"Criteria not defined for feature '{feature}'.")
+
+        return criteria
+
+    def get_bits_description(self, *, block: Block) -> dict[Feature, tuple[Criteria, FeatureState]]:
```

Comment:
Maybe create a namedtuple `FeatureDescription`?

Reply:
Done in https://github.com/HathorNetwork/hathor-core/pull/613/commits/ff0735143eceb321f880c207bb6544b138ac58e8

Reply:
Thanks!